examples/benchmarks/contextvalues_test.go

```diff
@@ -0,0 +1,139 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package benchmarks
+
+import (
+	"context"
+	"testing"
+
+	"github.com/go-logr/logr"
+	"github.com/go-logr/logr/funcr"
+)
+
+const iterationsPerOp = 100
+
+// 1% of the Info calls are invoked, all of those call the LogSink.
+func BenchmarkNewContext1Percent(b *testing.B) {
+	expectedCalls := int64(iterationsPerOp) / 100 * int64(b.N)
+	ctx := setup(b, expectedCalls)
+
+	// Each iteration is expected to do exactly the same thing,
+	// in particular do the same number of allocs.
+	for i := 0; i < b.N; i++ {
+		// Therefore we repeat newContext a certain number of
+		// times. Individual repetitions are allowed to sometimes log
+		// and sometimes not, but the overall execution is the same for
+		// every outer loop iteration.
+		for j := 0; j < iterationsPerOp; j++ {
+			run(ctx, logSomeEntries(j, 100, 0))
+		}
+	}
+}
+
+// 100% of the Info calls are invoked, none of those call the LogSink.
+func BenchmarkNewContext100PercentDisabled(b *testing.B) {
+	expectedCalls := int64(0)
+	ctx := setup(b, expectedCalls)
+
+	for i := 0; i < b.N; i++ {
+		for j := 0; j < iterationsPerOp; j++ {
+			run(ctx, logSomeEntries(j, 1, 2 /* not logged by default by funcr */))
+		}
+	}
+}
+
+// 100% of the Info calls are invoked, all of those call the LogSink.
+func BenchmarkNewContext100Percent(b *testing.B) {
+	expectedCalls := int64(b.N) * iterationsPerOp
+	ctx := setup(b, expectedCalls)
+
+	for i := 0; i < b.N; i++ {
+		for j := 0; j < iterationsPerOp; j++ {
+			run(ctx, logSomeEntries(j, 1, 0))
+		}
+	}
+}
+
+// 100% of the Info calls are invoked, with 10 Info calls per NewContext calls.
+func BenchmarkNewContextMany(b *testing.B) {
+	callsPerOp := 10
+	expectedCalls := int64(b.N) * iterationsPerOp * int64(callsPerOp)
+	ctx := setup(b, expectedCalls)
+
+	for i := 0; i < b.N; i++ {
+		for j := 0; j < iterationsPerOp; j++ {
+			run(ctx, logMultipleTimes(callsPerOp))
+		}
+	}
+}
+
+type contextKey1 struct{}
+type contextKey2 struct{}
+
+func run(ctx context.Context, op func(ctx context.Context)) {
+	// This is the currently recommended way of adding a value to a context
+	// and ensuring that all future log calls include it.  Trace IDs might
+	// get handled like this.
+	logger := logr.FromContextOrDiscard(ctx)
+	logger = logger.WithValues("i", 1, "j", 2)
+	ctx = context.WithValue(ctx, contextKey1{}, 1)
+	ctx = context.WithValue(ctx, contextKey2{}, 2)
+	ctx = logr.NewContext(ctx, logger)
+	op(ctx)
+}
+
+func logSomeEntries(j, mod, v int) func(ctx context.Context) {
+	return func(ctx context.Context) {
+		if j%mod == 0 {
+			logger := logr.FromContextOrDiscard(ctx)
```

Comment:
Because it was simple and actually getting a discard logger would have been detected by the output check. But `loggerFromContextOrDie` is better - changed.